genmain/genmain.go

```diff
@@ -72,8 +117,50 @@ func (m *Main) Kill(reason error) {
 	defer m.l.Unlock()
 
 	log(nil, reason).Info("shutting down")
-	for _, c := range m.components {
-		c.Tomb().Kill(reason)
+	shutdownStart := time.Now()
+	deadline := time.After(m.shutdownDeadline)
+
+	ctx := context.Background()
+	alive := m.components
+	inverseDeps := inverseDependencies(m.components)
+
+	for len(alive) > 0 {
+		var killed []Component
+
+		for _, component := range alive {
+			if IsReadyToKill(inverseDeps, component) {
+				component.Tomb().Kill(reason)
+				killed = append(killed, component)
+			}
+		}
+
+		if !waitAny(killed, deadline) {
```

Comment:
Is this going to report all the `killed` components as "too long to shut down" even if a single one miss the deadline?

Reply:
Is will report for all the components that are still alive, which is the same behaviour as before.